neuralpp/experiments/experimental_inference/graph_computation.py

```diff
@@ -31,80 +29,60 @@ def __call__(self, item):
+        children = self.full_tree.children(node)
+        if not get_child_scores_from_cache:
+            for child in children:
+                self.compute_result_dict(child)
+        candidates = [self[child] for child in children if self[child] is not None]
+        return argmax(candidates, lambda node_value_pair: node_value_pair.expansion_value)
+
     def compute_result_dict(self, node):
-        if node not in self.partial_tree:
-            result = (node, self.expansion_value_fn(node, self.partial_tree, self.full_tree))
+        if node not in self.tree:
+            self[node] = Expansion(node, self.expansion_value_function(node, self.tree, self.full_tree))
         else:
-            children = self.full_tree.children(node)
-            if children:
-                for child in children:
-                    self.compute_result_dict(child)
-                result = argmax(
-                    (self[child] for child in children),
-                    lambda node_value_pair: node_value_pair[1],
-                )
-            else:
-                result = None
-        self[node] = result
-
-    def update_value(self, target_node, changed_child, child_value):
+            self[node] = self.__compute_single_node(node, False)
+
+    def update_value(self, target_node, child_value):
         ancestor_value = self.result_dict.get(id(target_node))
-        if child_value is None:
-            # Current node is finished expanding.
-            # Attempt to replace ancestors' expansion value with the highest available child
-            update_candidates = [
-                self[child] for child in self.full_tree.children(target_node)
-                if self[child] is not None
-            ]
-            if update_candidates:
-                child_value = argmax(
-                    update_candidates,
-                    lambda node_value_pair: node_value_pair[1],
-                )
+        assert(ancestor_value is not None)
+        if child_value is None or child_value.expansion_value <= ancestor_value.expansion_value:
+            self[target_node] = self.__compute_single_node(target_node, True)
+        else:
             self[target_node] = child_value
-        elif ancestor_value is not None and child_value[1] <= ancestor_value[1]:
-            # If the updated value at node is already higher than any other value in the ancestor's subtree,
-            # we don't need to look at any other children. Otherwise, we need to check these values:
-            update_candidates = [
-                self[child] for child in self.full_tree.children(target_node)
-                if self[child] is not None
-            ]
-            update_candidates.append(child_value)
-            child_value = argmax(
-                update_candidates,
-                lambda node_value_pair: node_value_pair[1],
-            )
-        self[target_node] = child_value
 
     def expand_partial_tree_and_recompute(self, expansion_root):
```

Comment:
Note that `TreeComputation.invalidate` does not assume this is a partial tree to be expanded. It just invalidates a value for whatever reason. This makes `TreeComputation` more generic and maybe useful for situation in which we don't have an expanding partial tree, but something else.

I think it would be good to bring the same lack of assumption to `ExpansionValueComputation`. Currently it is offering `expand_partial_tree_and_recompute` to expand the tree, but this requires the class to know that it is operating on a partial tree and to know what is the full tree and so on. Note that this method is the only place where `expand_partial_tree_and_recompute` effectively `full_tree`. In the other methods, it passes `full_tree` to `expansion_value_function`, but we could avoid that by having the expansion function be created for an already specific full tree, externally, and only receive a node and the (partial) tree as arguments. This would "free" `ExpansionValueComputation` from the assumption that it is operating on a partial tree, and it would not have to expand the tree. It would simply compute the expansion value.

This would have the added benefit of moving the job of expanding the tree to where I think it belongs more, which is  `AnytimeExactBeliefPropagation`, because that's where intuitively the action of expanding a partial tree for better approximations happens.

Reply:
Tried to rework it in [this commit](https://github.com/rodrigodesalvobraz/neuralpp/commit/00aec9e565c7dd86153d5293bac0929b7cd59241), let me know what you think! The main issue was that the expansion was working on an implicitly larger partial tree (the AEBP partial tree + boundaries). I introduced `PartialExpansionTree` to represent this tree explicitly and think it's neater on the computation side, but I'm not sure if it's ok to have so many different tree classes floating around.